### ISSUE TITLE
[POS] Staking bloats UTXO set by splitting coinstake UTXOs

### ISSUE THREAD

**Issue by Aprogiena** [stratisproject/StratisBitcoinFullNode#550]
Currently, if a staking node finds a block, the UTXO used for that (the one that satisfies the target) is split if it is greater than 200 coins. This leads to big stakers having 1000s of UTXOs, which is bad for many reasons, one of the major problems being CPU consumption because the staking node has to go over each UTXO every 16 seconds to check if any satisfies the target. 

There are some reasons why the splitting was implemented, so despite the fact that those reasons are no longer necessary for current state of networks, we might want to preserve it for sidechains bootstrapping.

I suggest we just change the limit not to be 200 coins, but rather 2 % of the wallet balance. That will make sure that the original reason is preserved (one staker on the network vs. maturity of 50 blocks won't halt the chain) while it would prevent bloating to the current degree. The actual condition should be 

`split happens if UTXO's value is greater than (100 / REQIURED_COIN_AGE_FOR_STAKING) %`

**Comment by dangershony** [comment]
I don't understand this logic
`split happens if UTXO's value is greater than (100 / REQIURED_COIN_AGE_FOR_STAKING) %`

2% sound good alternatively, we can also use hard limits (because that reflects in CPU power) say more then 500 outputs in the wallet we split anything under that will not split anymore.

**Comment by Aprogiena** [comment]
the logic is that on testnet the maturity is 10 blocks, on mainnet it is 50 blocks and on sidechain, it can be whatever blocks

the 100/thatnumber should cause that enough UTXOs are available even if there is only on staker on the whole network (sidechain bootstrap)

actually we should be looking for `greater than or equal to` condition and `thatnumber+1`

so for mainnet with maturity 50, we would be looking for 100 / 51 = 1.96 %, so if you have 10000 coins in the wallet then your UTXO gets split in case it is greater than or equal to 196 coins. as such, you won't ever have lower than 98 coin UTXOs, which means you are going to have something between 52 and 102 UTXOs

so if your maturity is 50 and you have 52 UTXOs, it is guaranteed that at least one UTXO is always alive (unlocked, ready for staking), so the network won't die

so to correct myself:

`split happens if UTXO's value is greater than or equal to (100 / RCAFS) % of your total wallet balance`

but thinking about it, this does not really work if you start with just 1 UTXO, so that consensus rule for sidechains should enforce the coin maturity only after some time. if you touch the coins improperly though, it could still fail (if you join UTXOs manually). so maybe the consensus on sidechain should check if there are enough UTXOs in the whole system and maybe require 10x the limit before the limit is enforced, something like that

however, this problem with chain halt is present with current behavior as well, so this suggestion does not make it worse, we just need separated solution for sidechain bootstrapping

**Comment by Aprogiena** [comment]
500 outputs is too many if the maturity is 50 - for that you just need to make sure at least 52 UTXOs are there, so 500 is overkill

**Comment by dangershony** [comment]
That's a very good explanation thanks, at first glance it does make sense to only keep the required amount of UTXO to have at last one output to stake, but yeah its too risky if a user by mistake joins outputs.

We can prevent joining if from looking at the UTXO db, but also relaying on very precise numbers is risky what if there are utxo available so we allow a user to join outputs but the available UTXO are not staking, and the staking outputs have been join then chain will halt (at least for a while).

I would say the splitting should happen up to a threshold but such a precise calculation is too risky while the current threshold (200 coins) is too low.

But the main reason for the split avoidance is performance of the wallet when the output count is too high, that's why my suggestion of a hard coded value (we know that staking on X outputs will be efficient on most modern computers (or raspberry pies) and we try to split up to such a limit (can be configurable) but must not be lower then RCAFS).

**Comment by Aprogiena** [comment]
I think the dynamic calculation is better because it allows you to create different conditions for sidechain and the code won't break after that.

What you say is true, but I think the solution for that is to modify the consensus rule (on sidechain, because main chain does not have this problem anymore due to distribution). Such a consensus rule would say that the coin maturity rule applies only if number of UTXOs is greater than 10x the coin maturity limit. So if it was 50, we would require at least 500 UTXOs to exist in the system. Again, this can be gamed, but this is not the point. The point is to prevent halting the sidechain by mistake. If the main creator of the sidechain is willing to halt it, he certainly can (simply stop staking at any time during bootstrap), so we should only care about mistakes. With the rule I've mentioned using 10x or 5x requirement, the mistake should be eliminated completely and the formula presented before would work very well. Note that the formula guarantees a single extra output to exist in EVERY wallet, but it does not limit the wallet to have more outputs (very likely there will be 50 % more outputs than the limit) and that is in EVERY wallet, which means that once the chain is joined by second entity, the whole thing starts working very nicely and once there are enough entities to reach the 10x/5x UTXO count limit, coin maturity rule will jump in and the chain can be considered as fully bootstrapped. The only question then is whether we should apply this requirement only once (i.e. once reached, coin maturity is enabled for staking forever) or if we should dynamically switch it off and on again if the condition is violated/satisfied again. I think that one time switch should be enough as we really want to just prevent mistakes.

**Comment by Aprogiena** [comment]
If the above rationale sounds good, then for THIS issue it means that we want to rewrite coinstake splitting/joining mechanism in the code to meet the formula

`split happens if UTXO's value is greater than or equal to (100 / RCAFS) % of your total wallet balance`

and then have another issue to implement special rule for sidechain consensus regarding activation of coin maturity requirement for coinstake.

**Comment by Aprogiena** [comment]
Alternative solution to whole thing: Completely disable splitting, which just bloats UTXO and have a dynamic switch on coinstake coin maturity rule as stated above. If this is implemented, the UTXO won't grow artificially because of the splits and it will just take more time for the coin maturity rule to jump in, which is not necessarily a bad thing. Whole system would be much cleaner with this.

**Comment by dangershony** [comment]
Ok good comments, perhaps we should understand to more depth the reason behind maturity, is it for staking reasons? (more fair distribution of stake) so coins that found a stake must wait for x time before they can find another block. or is it so coins must be locked to allow creation of blocks, or any other reason.

**Comment by Aprogiena** [comment]
There are two things actually, one is maturity, which is a term that is used for newly minted coins. In the above discussion, we use it sometimes quite poorly. Second is age of coins (aka depth).

So there are two conditions:
1) coins have to be mature
2) coins have to be deep enough

In POS mainnet, both are 50, in testnet both values are 10. So the maturity is the usual thing as in other crypto - incentivize miner to work for the good of the chain so that they can use their coins once they mature. So if it is 50, miner is incentivize to do good for 50 more blocks on this chain.

The coin age criterion is there for POS reasons, which is more what you say. However, it does not make that much sense because as we know, we can make 1000 small UTXOs and only one will be blocked by this if you mine a block. Compare this to 1 big UTXO, which is entirely blocked for the period if the block is mined. 

I would say that I don't really understand why the second rule is there at all, maybe some practical reasons in implementation. Maybe clarity of the design, I don't know.

As for my previous comments - when I was talking about maturity rule, often I meant both these rules together - because if you remove just one, the other would jump in and could still block the coins. But still, both suggestions I've given are valid if we consider switching off both these rules. The incentive for the miner to do good on the chain is not necessary during sidechain bootstrapping.

**Comment by dangershony** [comment]
Your last point is not entirely correct the value of the UTXO is calculated towards the `weightedTarget` of the stake, so a UTXO with more value has a higher chance to find a block.

Coin depth has a consensus value, we cannot just discard it.

I suggest to modify your algo slightly and split until a UTXO count is above a multiple of `RCAFS`
`split happens if UTXO's count is bellow or equal to (RCAFS*10)`

**Comment by Aprogiena** [comment]
We can discard it on sidechain. The whole problem we are solving here only appears on sidechain because main chain mainnet and testnet don't have this problem. On main chain, we only want to put a limit on the split that reflects the consensus rule. And my suggestion was to make a rule that would conform to both uses - main chain and side chain - with one design. 

Ten times multiplier is a lot. Currently my node was close to unable to mine on testnet with 1000 UTXOs. 10 times multiplier is just 500 UTXOs on main net. If someone is running this on slower machine, 500 UTXOs can be just too many.

That's the reason I suggested to just guarantee the absolute minimum (on main chain).

But now it feels like this splitting thing is maybe not needed at all, since we can fix sidechain bootstrapping in different way - remove the consensus rule until it is safe.

**Comment by dangershony** [comment]
Removing splitting all together feels very dangerous, what if nodes suddenly stop minting,  or a blockchain becomes less popular and folk stake less.

I agree 10 times might be too much we could make that configurable with upper and lower limits.
`CAFS split happens if UTXO's count is bellow or equal to (RCAFS*[config -> min=1, default =3])`

**Comment by Aprogiena** [comment]
I argue that if the creator of the sidechain, which is likely to be the only miner during its bootstrap period, stops mining, the chain will die. So it does not matter how many UTXOs that miner has. Even if it has 1000 UTXOs, if it stops, it stops. 

If we only enabled the coin age rule once UTXO set is large enough then it is unlikely that anything bad happens. It would happen only if everyone agreed on stopping. Then again, regardless of the number, the chain would die. The splitting does not add anything to this imho.

**Comment by dangershony** [comment]
`unlikely` is problematic.
I would not go for disabling splitting completely but definitely improving the algo.
Can you comment on the min/default multiplier?

**Comment by Aprogiena** [comment]
multiplier of 3 by default should be OK, it would mean 150 UTXOs on main chain mainnet, which after optimization should be very negligible to calculate over even on slow computer, it should be (RCAFS+1) * multiplier though

i don't think this value should be a part of configuration as a number - just enable or disable splitting (enable means 3x multiplier, disabled means no splitting)

**Comment by dangershony** [comment]
Ok fine with me.

Well if we already put a configuration then make it an integer -1 (negative) is disabled anything else will be a multiplier (maybe a max)

**Comment by Aprogiena** [comment]
+ we talked about getting rid of the code that joins UTXOs to coinstake, which currently does not work for some reason - see #554 for details

**Comment by Aprogiena** [comment]
This has been implemented fully.
